File: fix_imports.py
#!/usr/bin/env python3
"""
Script to fix missing imports in Rust files.
Adds:
- Mutex (from parking_lot or std::sync)
- sleep (from tokio::time)
- interval (from tokio::time)
"""

import re
import sys
from pathlib import Path

def has_parking_lot_import(content):
    """Check if file already uses parking_lot"""
    return 'use parking_lot::' in content or 'use parking_lot::{' in content

def needs_mutex_import(content):
    """Check if file uses Mutex but doesn't import it"""
    # Check for Mutex:: usage
    if 'Mutex::new(' not in content and 'Mutex<' not in content:
        return False

    # Check if already imported
    if 'use parking_lot::Mutex' in content or 'use std::sync::Mutex' in content:
        return False
    if 'use parking_lot::{' in content and 'Mutex' in content.split('use parking_lot::{')[1].split('}')[0]:
        return False
    if 'use std::sync::{' in content and 'Mutex' in content.split('use std::sync::{')[1].split('}')[0]:
        return False

    return True

def needs_sleep_import(content):
    """Check if file uses sleep but doesn't import it"""
    if 'sleep(' not in content:
        return False

    # Check if already imported
    if 'use tokio::time::sleep' in content:
        return False
    if 'use tokio::time::{' in content and 'sleep' in content.split('use tokio::time::{')[1].split('}')[0]:
        return False

    return True

def needs_interval_import(content):
    """Check if file uses interval but doesn't import it"""
    if 'interval(' not in content:
        return False

    # Check if already imported
    if 'use tokio::time::interval' in content:
        return False
    if 'use tokio::time::{' in content and 'interval' in content.split('use tokio::time::{')[1].split('}')[0]:
        return False

    return True

def find_import_section_end(lines):
    """Find the last line of the import section"""
    last_use_line = -1
    for i, line in enumerate(lines):
        stripped = line.strip()
        if stripped.startswith('use '):
            last_use_line = i
        elif last_use_line > -1 and stripped and not stripped.startswith('//') and not stripped.startswith('use '):
            # Found first non-use, non-comment line after imports
            return last_use_line

    return last_use_line if last_use_line > -1 else 0

def add_mutex_import(lines, use_parking_lot):
    """Add Mutex import to the appropriate line"""
    if use_parking_lot:
        # Try to add to existing parking_lot import
        for i, line in enumerate(lines):
            if line.strip().startswith('use parking_lot::RwLock;'):
                lines[i] = line.replace('use parking_lot::RwLock;', 'use parking_lot::{RwLock, Mutex};')
                return True
            elif line.strip().startswith('use parking_lot::'):
                # Add Mutex to existing single import
                import_item = line.strip()[17:].rstrip(';')
                lines[i] = f'use parking_lot::{{{import_item}, Mutex}};\n'
                return True
            elif 'use parking_lot::{' in line:
                # Add to existing multi-import
                parts = line.split('}')
                items = parts[0].split('{')[1]
                if 'Mutex' not in items:
                    lines[i] = parts[0] + ', Mutex}' + '}'. join(parts[1:])
                return True

        # No existing parking_lot import, add new one
        insert_pos = find_import_section_end(lines) + 1
        lines.insert(insert_pos, 'use parking_lot::Mutex;\n')
    else:
        # Use std::sync::Mutex
        for i, line in enumerate(lines):
            if 'use std::sync::{' in line and 'Arc' in line:
                parts = line.split('}')
                items = parts[0].split('{')[1]
                if 'Mutex' not in items:
                    lines[i] = parts[0] + ', Mutex}' + '}'.join(parts[1:])
                return True

        # Add new import
        insert_pos = find_import_section_end(lines) + 1
        lines.insert(insert_pos, 'use std::sync::Mutex;\n')

    return True

def add_sleep_import(lines):
    """Add sleep import from tokio::time"""
    for i, line in enumerate(lines):
        if 'use tokio::time::{' in line:
            parts = line.split('}')
            items = parts[0].split('{')[1]
            if 'sleep' not in items:
                lines[i] = parts[0] + ', sleep}' + '}'.join(parts[1:])
            return True

    # Add new import
    insert_pos = find_import_section_end(lines) + 1
    lines.insert(insert_pos, 'use tokio::time::sleep;\n')
    return True

def add_interval_import(lines):
    """Add interval import from tokio::time"""
    for i, line in enumerate(lines):
        if 'use tokio::time::{' in line:
            parts = line.split('}')
            items = parts[0].split('{')[1]
            if 'interval' not in items:
                lines[i] = parts[0] + ', interval}' + '}'.join(parts[1:])
            return True
        elif line.strip() == 'use tokio::time::sleep;':
            lines[i] = 'use tokio::time::{sleep, interval};\n'
            return True

    # Add new import
    insert_pos = find_import_section_end(lines) + 1
    lines.insert(insert_pos, 'use tokio::time::interval;\n')
    return True

def fix_file(file_path):
    """Fix imports in a single file"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()

        lines = content.split('\n')

        # Determine what needs to be added
        needs_mutex = needs_mutex_import(content)
        needs_sleep = needs_sleep_import(content)
        needs_interval = needs_interval_import(content)

        if not (needs_mutex or needs_sleep or needs_interval):
            return False, "No changes needed"

        changes = []

        # Add imports
        if needs_mutex:
            use_parking_lot = has_parking_lot_import(content)
            add_mutex_import(lines, use_parking_lot)
            changes.append(f"Mutex ({'parking_lot' if use_parking_lot else 'std::sync'})")

        if needs_sleep:
            add_sleep_import(lines)
            changes.append("sleep")

        if needs_interval:
            add_interval_import(lines)
            changes.append("interval")

        # Write back
        new_content = '\n'.join(lines)
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(new_content)

        return True, f"Added: {', '.join(changes)}"

    except Exception as e:
        return False, f"Error: {str(e)}"

def main():
    """Main function to process all Rust files"""
    src_dir = Path('src')

    if not src_dir.exists():
        print("Error: src directory not found")
        sys.exit(1)

    rust_files = list(src_dir.rglob('*.rs'))

    fixed_count = 0
    for file_path in rust_files:
        changed, message = fix_file(file_path)
        if changed:
            print(f"✓ {file_path}: {message}")
            fixed_count += 1

    print(f"\nFixed {fixed_count} files")

if __name__ == '__main__':
    main()

File: fix_imports.rs
use std::fs;
use std::path::Path;
use walkdir::WalkDir;

fn needs_mutex_import(content: &str) -> bool {
    (content.contains("Mutex::new(") || content.contains("Mutex<")) &&
    !(content.contains("use parking_lot::Mutex") ||
      content.contains("use std::sync::Mutex") ||
      (content.contains("use parking_lot::{") && content.contains("Mutex")) ||
      (content.contains("use std::sync::{") && content.contains("Mutex")))
}

fn needs_sleep_import(content: &str) -> bool {
    content.contains("sleep(") &&
    !(content.contains("use tokio::time::sleep") ||
      (content.contains("use tokio::time::{") && content.contains("sleep")))
}

fn needs_interval_import(content: &str) -> bool {
    content.contains("interval(") &&
    !(content.contains("use tokio::time::interval") ||
      (content.contains("use tokio::time::{") && content.contains("interval")))
}

fn has_parking_lot(content: &str) -> bool {
    content.contains("use parking_lot::")
}

fn fix_file(path: &Path) -> Result<bool, std::io::Error> {
    let content = fs::read_to_string(path)?;
    let mut lines: Vec<&str> = content.lines().collect();

    let needs_mutex = needs_mutex_import(&content);
    let needs_sleep = needs_sleep_import(&content);
    let needs_interval = needs_interval_import(&content);

    if !needs_mutex && !needs_sleep && !needs_interval {
        return Ok(false);
    }

    let mut changes = Vec::new();
    let mut modified = false;

    // Find the last `use` statement
    let mut last_use_idx = 0;
    for (i, line) in lines.iter().enumerate() {
        let trimmed = line.trim();
        if trimmed.starts_with("use ") {
            last_use_idx = i;
        }
    }

    // Fix Mutex import
    if needs_mutex {
        let use_parking_lot = has_parking_lot(&content);
        let mut mutex_added = false;

        for i in 0..lines.len() {
            let line = lines[i];

            if use_parking_lot {
                if line.trim() == "use parking_lot::RwLock;" {
                    lines[i] = "use parking_lot::{RwLock, Mutex};";
                    changes.push("Mutex (parking_lot)");
                    mutex_added = true;
                    modified = true;
                    break;
                } else if line.contains("use parking_lot::{") && !line.contains("Mutex") {
                    let new_line = line.replace("}", ", Mutex}");
                    lines[i] = Box::leak(new_line.into_boxed_str());
                    changes.push("Mutex (parking_lot)");
                    mutex_added = true;
                    modified = true;
                    break;
                }
            } else {
                if line.contains("use std::sync::{") && line.contains("Arc") && !line.contains("Mutex") {
                    let new_line = line.replace("}", ", Mutex}");
                    lines[i] = Box::leak(new_line.into_boxed_str());
                    changes.push("Mutex (std::sync)");
                    mutex_added = true;
                    modified = true;
                    break;
                }
            }
        }

        if !mutex_added {
            let import_line = if use_parking_lot {
                "use parking_lot::Mutex;"
            } else {
                "use std::sync::Mutex;"
            };
            lines.insert(last_use_idx + 1, Box::leak(import_line.to_string().into_boxed_str()));
            changes.push(if use_parking_lot { "Mutex (parking_lot)" } else { "Mutex (std::sync)" });
            modified = true;
        }
    }

    // Fix sleep import
    if needs_sleep {
        let mut sleep_added = false;

        for i in 0..lines.len() {
            let line = lines[i];

            if line.contains("use tokio::time::{") && !line.contains("sleep") {
                let new_line = line.replace("}", ", sleep}");
                lines[i] = Box::leak(new_line.into_boxed_str());
                changes.push("sleep");
                sleep_added = true;
                modified = true;
                break;
            }
        }

        if !sleep_added {
            lines.insert(last_use_idx + 1, "use tokio::time::sleep;");
            changes.push("sleep");
            modified = true;
        }
    }

    // Fix interval import
    if needs_interval {
        let mut interval_added = false;

        for i in 0..lines.len() {
            let line = lines[i];

            if line.contains("use tokio::time::{") && !line.contains("interval") {
                let new_line = line.replace("}", ", interval}");
                lines[i] = Box::leak(new_line.into_boxed_str());
                changes.push("interval");
                interval_added = true;
                modified = true;
                break;
            } else if line.trim() == "use tokio::time::sleep;" {
                lines[i] = "use tokio::time::{sleep, interval};";
                changes.push("interval");
                interval_added = true;
                modified = true;
                break;
            }
        }

        if !interval_added {
            lines.insert(last_use_idx + 1, "use tokio::time::interval;");
            changes.push("interval");
            modified = true;
        }
    }

    if modified {
        let new_content = lines.join("\n") + "\n";
        fs::write(path, new_content)?;
        println!("✓ {}: Added {}", path.display(), changes.join(", "));
    }

    Ok(modified)
}

fn main() {
    let mut fixed_count = 0;

    for entry in WalkDir::new("src").follow_links(true) {
        let entry = match entry {
            Ok(e) => e,
            Err(e) => {
                eprintln!("Error walking directory: {}", e);
                continue;
            }
        };

        let path = entry.path();

        if path.extension().and_then(|s| s.to_str()) == Some("rs") {
            match fix_file(path) {
                Ok(true) => fixed_count += 1,
                Ok(false) => {},
                Err(e) => eprintln!("Error processing {}: {}", path.display(), e),
            }
        }
    }

    println!("\nFixed {} files", fixed_count);
}
